### PR TITLE
GFM compliance for tasks lists

### LINKED ISSUE #107: To do lists
https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments

To be honest, I'm not entirely sure if Markdown should support this or not. Seems very, very Github centric, but curious what your thoughts are.

### LINKED ISSUE #170: Enable task lists (from GFM)
GFM lets you create [Task Lists](https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments) using the following syntax:

```
- [ ] First item
- [x] Checked item
```

Which renders the following:
- [ ] First item
- [x] Checked item

It would be great to have that feature in marked.

### LINKED ISSUE #419: support of task lists from github flavored markdown
Do you have plan to add the support of task lists from github flavored markdown?
https://github.com/blog/1375%0A-task-lists-in-gfm-issues-pulls-comments

### LINKED ISSUE #689: GFM: my task lists not rendered
my test code:

``` markdown
### Solar System Exploration, 1950s – 1960s

- [ ] Mercury
- [x] Venus
- [x] Earth (Orbit/Moon)
- [x] Mars
- [ ] Jupiter
- [ ] Saturn
- [ ] Uranus
- [ ] Neptune
- [ ] Comet Haley
```

results:

```
Solar System Exploration, 1950s – 1960s

- [ ] Mercury
- [x] Venus
- [x] Earth (Orbit/Moon)
- [x] Mars
- [ ] Jupiter
- [ ] Saturn
- [ ] Uranus
- [ ] Neptune
- [ ] Comet Haley
```

my config:

``` javascript
marked.setOptions({
    gfm: true,
    tables: true,
    highlight: function (code, lang) {
        return hljs.highlightAuto(code, [lang]).value;
    }
});
```

### PR DESCRIPTION
**Pull request by tomtheisen**
**Marked version:**  0.3.19

**Markdown flavor:**  GitHub Flavored Markdown

Fixes #107
Fixes #419
Fixes #170
Fixes #689
Fixes #587

## Description

This adds [task list items](https://github.github.com/gfm/#task-list-items-extension-) support as specified in GFM.  These are the checkboxes you can make in bulleted lists.

I did do some funny stuff in the gfm test runner.  In particular all the other gfm tests use the xhtml rendering option.  But for some reason, (probably not a good one) the checkbox rendering in the gfm spec is *not* xhtml.  So, I added a an `xhtml` attribute to the test spec json for those tests.  It seems pretty janky.  Maybe someone can think of a better way of handling it.

## Contributor

- [X] Test(s) exist to ensure functionality and minimize regression (if no tests added, list tests covering this PR); or,
- [ ] If submitting new feature, it has been documented in the appropriate places.

## Committer

In most cases, this should be a different person than the contributor.

- [x] Draft GitHub release notes have been updated.
- [x] CI is green (no forced merge required).
- [x] Merge PR

### CODE REVIEW

File: lib/marked.js [hunk 1]
```diff
@@ -927,6 +939,14 @@ Renderer.prototype.listitem = function(text) {
   return '<li>' + text + '</li>\n';
 };
 
+Renderer.prototype.checkbox = function(checked) {
+  return '<input '
+    + (checked ? 'checked="" ' : '')
+    + 'disabled="" type="checkbox"'
+    + (this.options.xhtml ? ' /' : '')
+    + '>';
```
Comment (UziTech) [inline lib/marked.js]: from the specs it looks like there should be a space after `>`

File: test/specs/gfm/gfm-spec.js [hunk 2]
```diff
@@ -15,7 +15,8 @@ Messenger.prototype.test = function(spec, section, ignore) {
     var shouldFail = ~ignore.indexOf(spec.example);
     it('should ' + (shouldFail ? 'fail' : 'pass') + ' example ' + spec.example, function() {
       var expected = spec.html;
-      var actual = marked(spec.markdown, { headerIds: false, xhtml: true });
+      var usexhtml = typeof spec.xhtml === 'boolean' ? spec.xhtml : true;
+      var actual = marked(spec.markdown, { headerIds: false, xhtml: usexhtml });
```
Comment (UziTech) [inline test/specs/gfm/gfm-spec.js]: It looks like just setting the `xhtml` option to false for all gfm the tests works

File: lib/marked.js [hunk 3]
```diff
@@ -363,10 +365,20 @@ Lexer.prototype.token = function(src, top) {
           if (!loose) loose = next;
         }
 
+        // Check for task list items
+        istask = /^\[[ xX]\]/.test(item);
```
Comment (UziTech) [inline lib/marked.js]: a space seems to be required after `[ ]` other wise it is rendered like text

with space:

- [x] test

without space:

- [x]test

File: lib/marked.js [hunk 4]
```diff
@@ -363,10 +365,20 @@ Lexer.prototype.token = function(src, top) {
...
+        ischecked = undefined;
+        if (istask) {
+          ischecked = item[1] !== ' ';
+          item = item.replace(/^\[[ xX]\] */, '');
```
Comment (UziTech) [inline lib/marked.js]:
```diff
- /^\[[ xX]\] */
+ /^\[[ xX]\] +/
```
since at least one space is required

File: test/specs/gfm/gfm-spec.js (hunk 2 above)
Comment (joshbruce) [reply test/specs/gfm/gfm-spec.js]: I think CommonMark uses XHTML maybe the GFM extensions don't need XHTML. (I do think there's something to be said for the attempted solution - putting the options in the JSON - but not sure we've hit a point of actually needing it.)

**Comment by tomtheisen** [comment]
I've addressed a few things.

1.  xhtml option is hardcoded to false for gfm tests.
2.  Space is required following task markdown.
3.  Task checkbox html is rendered with a space following it.

**Comment by tomtheisen** [comment]
I'm on vacation this week, and I resolved not to work on work, but I have to work on something!

**Comment by styfle** [comment]
@tomtheisen This is huge! 🏋️ 
Thanks for implementing this! 🎉

**Pull request merged**